### ISSUE TITLE
Memory Leak in Tests

### ISSUE THREAD

**Issue by jamesarosen** [ember-cli/ember-ajax#104]
I'm converting an Ember 1.13 app from ember-cli-ic-ajax to ember-ajax and my tests keep running out of memory. See the discussion below for the (likely) cause.

**Comment by jamesarosen** [comment]
Retainers:
```
container in Class
  __proto__ in Class
    service:ajax in @17562093
      cache in Container
        container in Class
          __proto__ in Class
            _this in system / Context
              context in function() (ajax-request.js:53)
                [1] in Array
                  [13] in Array
                    waiters in @16319955
                      Test in @16339013
                        Em in window
```

**Comment by alexlafroscia** [comment]
Hey @jamesarosen! Thanks for reporting this.  I've made a PR to remove the `options.context = this` bit that seems to be causing problems; it didn't seem to be necessary anyway.  Would you mind giving that branch a try and letting me know if it fixes things for you?

**Comment by jamesarosen** [comment]
Thanks for the quick turnaround. Your proposed fix doesn't quite do it because even though `options.context` is no longer a pointer the `service:ajax`, `_this` is. There's no way to get around the fact that those waiters have a reference to the `service:ajax`. Could you maybe store `pendingRequestCount` on some other object that _doesn't_ have a reference to `service:ajax`?

**Comment by jamesarosen** [comment]
You might be able to do it with a closure-local variable:

```js
let pendingRequestCount = 0;

export default class AjaxRequest {
  // in case other clients of service:ajax need access to this variable:
  get pendingRequestCount() {
    return pendingRequestCount;
  }

  init() {
    pendingRequestCount = 0;
    if (testing) {
      Test.registerWaiter(() => pendingRequestCount === 0);
    }
  }
}
```
Then the callback only has a handle on the _module_, which the requireJS global module system already has a handle on, so there's no new leak.

**Comment by alexlafroscia** [comment]
I don't think that would matter (per-instance of `service:ajax` vs. global) but I'd like to ping @rwjblue on that to ask, since this implementation is based on his suggestion.  

Also, thanks for the patience and good explanations; I've never dealt much with memory issues so this is a huge learning opportunity for me!

**Comment by alexlafroscia** [comment]
Also, just to I can understand _why_ this was an issue, from a learning-more-about-JS perspective...

The issue is that `Test.registerWaiter` has a reference to an instance of the service, and isn't letting go of that after each test finishes.  So, each test is generating at least one reference to the service that never is actually released, and it eventually becomes too much for the browser to handle.

Is that right?

**Comment by jamesarosen** [comment]
> So, each test is generating at least one reference to the service that never is actually released, and it eventually becomes too much for the browser to handle.

Exactly! Though it's not the `service:ajax` that's the memory hog. It's the fact that the service is built from the container and thus has a reference to the container. That, in turn, has a reference to every route, controller, template, and service in the app. Our app has about 1000 acceptance tests. If we leak a `container` per test, we run into the many GB of memory usage.

**Comment by alexlafroscia** [comment]
Oh wow, I see. So, if it were just the ES6 class it wouldn't be (as much of) a problem, right?

**Comment by jamesarosen** [comment]
> So, if it were just the ES6 class it wouldn't be (as much of) a problem

Probably not too bad, no. Still a leak, but most systems probably wouldn't notice.

**Comment by alexlafroscia** [comment]
Cool.  If we used `unregisterWaiter`, do you think we avoid the issue of the container not being destroyed with each test?  Maybe we could keep the per-instance counter, but still avoid the leak.

**Comment by jamesarosen** [comment]
> If we used `unregisterWaiter`, do you think we avoid the issue of the container not being destroyed with each test?

It may. I don't know much about how `Test.registerWaiter` works. I'd be happy to try that variant as well.

**Comment by alexlafroscia** [comment]
@jamesarosen awesome, thanks. I pushed a commit that moves back to the per-instance count, and then tears down the waiter when the service is destroyed (which should be done at the end of each acceptance test).

**Comment by jamesarosen** [comment]
Now I see the leak again, though I can no longer tie it directly to this path, so I can't really say whether the `unregisterWaiter` method works.

**Comment by alexlafroscia** [comment]
Thanks for working with me on this @jamesarosen! I just released `v2.3.2` that includes this fix.
